include/xtensor/xfunction.hpp

```diff
@@ -540,14 +556,54 @@ namespace xt
     template <class F, class R, class... CT>
     inline auto xfunction_base<F, R, CT...>::dimension() const noexcept -> size_type
     {
-        size_type dimension = m_shape_computed ? m_shape.size() : compute_dimension();
-        return dimension;
+        return xtl::mpl::static_if<no_broadcast>([&](auto self)
+        {
+            XTENSOR_ASSERT(std::get<0>(self(*this).m_e).dimension() == self(*this).compute_dimension());
+            return std::get<0>(self(*this).m_e).dimension();
+        },
+        /*else*/
+        [&](auto self)
+        {
+            return self(*this).m_shape_computed ? self(*this).m_shape.size() : self(*this).compute_dimension();
+        });
     }
 
     /**
      * Returns the shape of the xfunction.
      */
     template <class F, class R, class... CT>
+    template <bool NB, std::enable_if_t<NB, int>>
+    inline auto xfunction_base<F, R, CT...>::shape() const -> const shape_type&
+    {
+        if (!m_shape_computed)
+        {
+            std::copy(std::get<0>(m_e).shape().begin(), std::get<0>(m_e).shape().end(), m_shape.begin());
+            XTENSOR_TRY(broadcast_shape(m_shape));  // make sure broadcast works in debug mode
+            m_shape_computed = true;
+        }
+        return m_shape;
+
+        // The code below claims to return a reference to a temporary
+        // **********************************************************
+        // using first_shape = typename std::decay_t<detail::pack_first_t<CT...>>::inner_shape_type;
+        // return xtl::mpl::static_if<std::is_same<shape_type, first_shape>::value>(
+        //     [&](auto self)
+        //     {
+        //         return std::get<0>(self(*this).m_e).shape();
+        //     },
+        //     [&](auto self)
+        //     {
+        //         if (!self(*this).m_shape_computed)
+        //         {
+        //             std::copy(std::get<0>(self(*this).m_e).shape().begin(), std::get<0>(self(*this).m_e).shape().end(), self(*this).m_shape.begin());
+        //         }
+        //         return self(*this).m_shape;
+        //     }
+        // );
+    }
```

Comment:
We need to investigate this once the PR is merged, having a single method gives a better API.